### PR TITLE
Extended ps

### PR DESCRIPTION
**Pull request by ajayparida**
Listen interval parameters from user/app

### CODE REVIEW

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 1]
```diff
@@ -4392,3 +4392,41 @@ enum wifi_nrf_status wifi_nrf_fmac_set_listen_interval(void *dev_ctx,
...
+							bool ps_wake_listen_interval)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct nrf_wifi_umac_cmd_config_extended_ps *set_ps_wake_interval_cmd = NULL;
+	struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx = NULL;
+
+	fmac_dev_ctx = dev_ctx;
+
+	set_ps_wake_interval_cmd = wifi_nrf_osal_mem_zalloc(fmac_dev_ctx->fpriv->opriv,
+						       sizeof(*set_ps_wake_interval_cmd));
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c]: Indentation

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c [hunk 2]
```diff
@@ -825,3 +825,43 @@ int wifi_nrf_set_listen_interval(const struct device *dev,
...
+
+int wifi_nrf_set_ps_wake_interval(const struct device *dev,
+				  struct wifi_ps_wake_interval_params *params)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct wifi_nrf_ctx_zep *rpu_ctx_zep = NULL;
+	struct wifi_nrf_vif_ctx_zep *vif_ctx_zep = NULL;
+	int ret = -1;
+
+	if (!dev || !params) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c]: Add an error message

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c [hunk 3]
```diff
@@ -825,3 +825,43 @@ int wifi_nrf_set_listen_interval(const struct device *dev,
...
+		LOG_ERR("%s: rpu_ctx_zep is NULL\n", __func__);
+		goto out;
+	}
+
+	status = wifi_nrf_fmac_set_ps_wake_interval(rpu_ctx_zep->rpu_ctx,
+						    vif_ctx_zep->vif_idx,
+						    params->ps_wake_interval);
+
+	if (status != WIFI_NRF_STATUS_SUCCESS) {
+		LOG_ERR("%s: wifi_nrf_fmac_ failed\n", __func__);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c]: wifi_nrf_fmac_ should be wifi_nrf_fmac_set_ps_wake_interval

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 4]
```diff
@@ -4394,38 +4394,38 @@ enum wifi_nrf_status wifi_nrf_fmac_set_listen_interval(void *dev_ctx,
 }
 
 
-enum wifi_nrf_status wifi_nrf_fmac_set_ps_wake_interval(void *dev_ctx,
-							unsigned char if_idx,
-							bool ps_wake_listen_interval)
+enum wifi_nrf_status wifi_nrf_fmac_set_ps_wakeup_mode(void *dev_ctx,
+				      unsigned char if_idx,
+				      bool ps_wakeup_mode)
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:4399]: Indentation

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c [hunk 5]
```diff
@@ -217,63 +218,107 @@ int wifi_nrf_set_power_save(const struct device *dev,
...
+							rpu_ctx_zep->rpu_ctx,
+							vif_ctx_zep->vif_idx,
+							UAPSD_Q_MAX);
+		} else {
+			/* Legacy mode*/
+			status = wifi_nrf_fmac_set_uapsd_queue(
+							rpu_ctx_zep->rpu_ctx,
+							vif_ctx_zep->vif_idx,
+							UAPSD_Q_MIN);
+		}
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c:279]: Using an intermediate variable a single invocation should suffice.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 6]
```diff
@@ -1406,4 +1406,33 @@ enum wifi_nrf_status wifi_nrf_fmac_get_host_rpu_ps_ctrl_state(void *fmac_dev_ctx
...
+
+/**
+ * @brief Configure WLAN ps wakeup mode to DTIM or listen interval.
+ * @param[in] fmac_dev_ctx Pointer to the UMAC IF context for a RPU WLAN device.
+ * @param[in] if_idx Index of the interface on which power management is to be set.
+ * @param[in] ps_wakeup_mode Enable listen interval based ps wakeup.
+ *
+ * @details This function is used to send a command to RPU to set ps wakeup mode
+ *	    to listen interval. Default ps wakeup mode DTIM.
+ *
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1432]:
```suggestion
 * @details This function is used to configure PS wakeup mode, either DTIM Interval or Lister interval based . Default is DTIM interval based mode.
 *
```

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 7]
```diff
@@ -1406,4 +1406,33 @@ enum wifi_nrf_status wifi_nrf_fmac_get_host_rpu_ps_ctrl_state(void *fmac_dev_ctx
...
+ * @details This function is used to send a command to RPU to configure listen interval.
+ *
+ * @return #wifi_nrf_status
+ */
+enum wifi_nrf_status wifi_nrf_fmac_set_listen_interval(void *fmac_dev_ctx,
+						       unsigned char if_idx,
+						       unsigned short listen_interval);
+
+/**
+ * @brief Configure WLAN ps wakeup mode to DTIM or listen interval.
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1425]:
```suggestion
 * @brief Configure WLAN PS wakeup mode to DTIM interval or listen interval.
```

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw/host_rpu_umac_if.h [hunk 8]
```diff
@@ -3212,6 +3212,11 @@ struct nrf_wifi_umac_cmd_config_extended_ps {
 #define NRF_WIFI_MAX_TWT_FLOWS 8
 #define NRF_WIFI_PS_MODE_LEGACY 0
 #define NRF_WIFI_PS_MODE_WMM 1
+/* As per hardware design for power efficiency
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fw/host_rpu_umac_if.h:3215]: I don't think this is has to be with hardware design, but I might be wrong, can we elaborate on this? The comment should explain why 3, it now just explains that it is 3.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 9]
```diff
@@ -1420,4 +1420,19 @@ enum wifi_nrf_status wifi_nrf_fmac_get_host_rpu_ps_ctrl_state(void *fmac_dev_ctx
 enum wifi_nrf_status wifi_nrf_fmac_set_listen_interval(void *fmac_dev_ctx,
 						       unsigned char if_idx,
 						       unsigned short listen_interval);
+
+/**
+ * @brief Configure WLAN ps to dtim or listen interval.
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h]: wrong description

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 10]
```diff
@@ -1420,4 +1420,19 @@ enum wifi_nrf_status wifi_nrf_fmac_get_host_rpu_ps_ctrl_state(void *fmac_dev_ctx
...
+ * @param[in] fmac_dev_ctx Pointer to the UMAC IF context for a RPU WLAN device.
+ * @param[in] if_idx Index of the interface on which power management is to be set.
+ * @param[in] ps_wake_interval Enable listen interval based ps(default dtim based)
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h]: Description looks wrong, we now have a separate mode, so, this should just be interval, no?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c [hunk 11]
```diff
@@ -217,63 +218,107 @@ int wifi_nrf_set_power_save(const struct device *dev,
...
-	if (!rpu_ctx_zep) {
-		LOG_ERR("%s: rpu_ctx_zep is NULL\n", __func__);
-		goto out;
-	}
+		status = wifi_nrf_fmac_set_power_save_timeout(
+							rpu_ctx_zep->rpu_ctx,
+							vif_ctx_zep->vif_idx,
+							params->timeout_ms);
+
+		if (status != WIFI_NRF_STATUS_SUCCESS) {
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c:258]: All status checks can be moved out of switch, we can just dump `%s: Configuring %d failed, __func__, params->type`, makes code cleaner

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c [hunk 12]
```diff
@@ -739,129 +787,3 @@ void wifi_nrf_event_proc_twt_sleep_zep(void *vif_ctx,
...
-		LOG_ERR("%s: wifi_nrf_fmac_ failed\n", __func__);
-		goto out;
-	}
-
-	ret = 0;
-out:
-	return ret;
-}
-
-int wifi_nrf_set_ps_wake_interval(const struct device *dev,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wifi_mgmt.c]: This is a fixup commit, as this function changes in previous commit, so, either fix or squash/fixup these two commits.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 13]
```diff
@@ -1414,25 +1414,25 @@ enum wifi_nrf_status wifi_nrf_fmac_get_host_rpu_ps_ctrl_state(void *fmac_dev_ctx
  * @params[in] listen_intvl listen interval to be configured.
  *
  * @details This function is used to send a command to RPU to configure listen interval.
- *
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h]: unnecessary whitespace

**Review by krish2718 — CHANGES_REQUESTED** [review]
* PS is an acronym, so, should be all caps.
* the commit log says `set as received from f/w` but should be ` set as received from Zephyr` ?

**Review by sachinthegreen — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

File: west.yml [hunk 14]
```diff
@@ -59,7 +59,7 @@ manifest:
     # https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/modules.html
     - name: zephyr
       repo-path: sdk-zephyr
-      revision: 8182ebc2d257c309b9c50fc0d895a4b88a672a1b
+      revision: pull/1162/head
```
Comment (rlubos) [inline west.yml:62]: Let's do the final manifest update here (`dfd1fad51fa591b44ed44159e54a97625d317a99`).

Just need to wait for https://github.com/nrfconnect/sdk-nrf/pull/11109 as it'll trigger conflict.

**Comment by rlubos** [comment]
https://github.com/nrfconnect/sdk-nrf/pull/11109 was just merged, please rebase/update manfiest

**Comment by rlubos** [comment]
The sdk-zephyr SHA should still be updated to `dfd1fad51fa591b44ed44159e54a97625d317a99`, otherwise it won't pick up the latest merged changes (from https://github.com/nrfconnect/sdk-zephyr/pull/1162)

**Comment by krish2718** [comment]
> The sdk-zephyr SHA should still be updated to `dfd1fad51fa591b44ed44159e54a97625d317a99`, otherwise it won't pick up the latest merged changes (from [nrfconnect/sdk-zephyr#1162](https://github.com/nrfconnect/sdk-zephyr/pull/1162))

I have pushed now, consolidated all `manifest-zephyr` PRs to this one.

**Review by rlubos — APPROVED** [review]

**Comment by rlubos** [comment]
Merging, as it was agreed with @nordic-hani to merge this post RC1 tag.

**Pull request merged**